lua/neogit/buffers/fuzzy_finder.lua

```diff
@@ -0,0 +1,80 @@
+local Finder = require("neogit.lib.finder")
+
+local function buffer_height(count)
+  if count < (vim.fn.winheight(0) / 2) then
+    return count + 2
+  else
+    return 0.5
+  end
+end
+
+---@class FuzzyFinderBuffer
+---@field list table list of items to search
+---@field action function action dispatched by line selection
+---@field buffer Buffer
+---@field open_sync function
+---@field open function
+local M = {}
+
+---Creates a new FuzzyFinderBuffer
+---@param list any[]
+---@return FuzzyFinderBuffer
+function M.new(list)
+  local instance = {
+    list = list,
+  }
+
+  setmetatable(instance, { __index = M })
+
+  return instance
+end
+
+function M:open(opts, action)
+  opts = opts or {
+    allow_multi = false,
+    layout_config = { height = buffer_height(#self.list) },
+  }
+
+  Finder.create(opts):add_entries(self.list):find(action)
+end
+
+-- Opens finder in such a way that selected value can be returned to the main thread
+-- without the need to use a callback to process the selection.
+function M:open_sync(...)
```

Comment:
You may have other opinions on this, but from what I've seen `sync` refers to the whole editor blocking on the call, like a prompt using the older style vimscript function like `confirm()` or other ex line prompts. Libuv uses this notation for example, where `sync` methods just freeze the whole editor in place until it is done, which is generally not preferred.

Whereas `async` only suspends the current task and keeps all other things turning.

The plenary async tasks `async.run` `async.void` and everything run inside of them spawn a lua coroutine and when awaiting for a result the coroutine is suspended, and resumed when a value is available, which is what neogit uses as much as possible.

The third method is callbacks, which is the style that vim.loop (https://github.com/luvit/luv) uses.

These can be adapted to the coroutine/task/green thread style using `plenary.async.wrap`, which is how `find_async` uses vim.ui.select which is callback style.

I do not know which language you are the most familiar to, but the plenary async model is quite similar to go's async model.

Let me know if you have any opinions on the naming

Reply:
I think I just had a misunderstanding of how sync/async/blocking nomenclature worked. I mostly write ruby at the dayjob, so the only real concurrency I've dealt with is... pretty much none, actually. Which is also to say that I appreciate the pointers ;) 

So, if it makes more sense to rename the method, then I'm all for it.